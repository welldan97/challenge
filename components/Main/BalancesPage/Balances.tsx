import React, { memo } from 'react';
import styled from 'styled-components';

// SECTION: Styles

const Table = styled.table`
  margin-left: 24px;
  margin-right: 24px;
`;

const Row = styled.tr`
  height: 75px;
  & + & {
    border-top: 1px solid #e0e0f4;
  }
`;

const Address = styled.span`
  display: block;
  font-size: 13px;
  letter-spacing: 0.05em;
  transform: translateY(3px);
`;

const ValueCell = styled.td`
  text-align: right;
  transform: translateY(1px);
`;

const Value = styled.span`
  font-size: 16px;
  letter-spacing: 0.05em;
  text-align: right;
  font-weight: bold;
`;

const Symbol = styled.span`
  font-size: 12px;
  letter-spacing: 0.05em;
  text-align: right;
`;

// !SECTION
// SECTION: Subcomponents

interface BalanceProps {
  balance: {
    address: string;
    value: string;
  };
}

const Balance = memo(({ balance }: BalanceProps) => (
  <Row>
    <td>
      <Address>{balance.address}</Address>
    </td>
    <ValueCell>
      <Value>{balance.value}</Value>
      <Symbol> ETH</Symbol>
    </ValueCell>
  </Row>
));

// !SECTION
// SECTION: Main

const Addresses = memo(() => {
  const balances = [
    {
      address: '0x32be343b94f860124dc4fee278fdcbd38c102d88',
      value: '0.04793',
    },
    {
      address: '0x32be343b94f860924dc4fee278fdcbd38c102d88',
      value: '0.04793',
    },
    {
      address: '0x32be343b94f860824dc4fee278fdcbd38c102d88',
      value: '0.04793',
    },
    {
      address: '0x32be343b94f860224dc4fee278fdcbd38c102d88',
      value: '0.04793',
    },
  ];

  return (
    <Table>
      <tbody>
        {balances.map(balance => (
          <Balance key={balance.address} balance={balance} />
        ))}
      </tbody>
    </Table>
  );
});

export default Addresses;
