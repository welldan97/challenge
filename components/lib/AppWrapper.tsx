import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode, Suspense } from 'react';
import ErrorBoundary from '../../lib/ErrorBoundary';

// SECTION: Main

interface Props {
  children: ReactNode;
}

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
    },
  },
});

export default function AppWrapper({ children }: Props) {
  return (
    <ErrorBoundary fallback={<h1>Error</h1>}>
      <Suspense fallback={<h1>Loading...</h1>}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Suspense>
    </ErrorBoundary>
  );
}
