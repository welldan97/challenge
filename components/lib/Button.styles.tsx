import styled, { css } from 'styled-components';

// SECTION: Main

const styles = css`
  background: #6932d4;
  border-radius: 100px;
  color: white;
  border: none;
  height: 40px;
  width: 88px;
  appearance: none;
  text-decoration: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  &:hover,
  &:focus,
  &:active {
    color: white;
  }
`;

export const StyledButton = styled.button`
  ${styles}
`;
export const StyledAnchor = styled.a`
  ${styles}
`;
