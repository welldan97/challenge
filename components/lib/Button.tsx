import omit from 'lodash/fp/omit';
import Link from 'next/link';
import { ComponentPropsWithoutRef, memo } from 'react';
import styled, { css } from 'styled-components';

// SECTION: Styles

const styles = css`
  background: #6932d4;
  border-radius: 100px;
  color: white;
  border: none;
  height: 40px;
  width: 88px;
  appearance: none;
  text-decoration: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  &:hover,
  &:focus,
  &:active {
    color: white;
  }
`;

export const StyledButton = styled.button`
  ${styles}
`;
export const StyledAnchor = styled.a`
  ${styles}
`;

// !SECTION
// SECTION: Main

interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
  type?: ComponentPropsWithoutRef<'button'>['type'];
}

interface AnchorProps extends ComponentPropsWithoutRef<'a'> {
  type: 'anchor';
}

const Button = memo((props: ButtonProps | AnchorProps = { type: 'button' }) => {
  if (props.type === 'anchor') {
    const passedHref = typeof props.href === 'string' ? props.href : '';
    return (
      <Link href={passedHref}>
        <StyledAnchor {...omit(['type'], props)} />
      </Link>
    );
  }

  return <StyledButton {...props} />;
});

export default Button;
