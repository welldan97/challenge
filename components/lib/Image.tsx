import React, { memo } from 'react';
import BaseImage, { ImageProps } from 'next/image';

// SECTION: Main

const Image = memo(({ className, style, ...props }: ImageProps) => (
  <div className={className} style={style}>
    <BaseImage {...props} />
  </div>
));

export default Image;
