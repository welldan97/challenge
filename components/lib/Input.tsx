import { useField } from 'formik';
import React, { memo } from 'react';
import styled from 'styled-components';

// SECTION: Styles

const Wrapper = styled.div`
  padding-left: 24px;
  padding-right: 24px;
  height: 88px;
`;

const Label = styled.label`
  display: block;
  margin-bottom: 6px;
  margin-left: 8px;
`;

const BaseInput = styled.input`
  height: 44px;
  background: #ffffff;
  border: 1px solid #6932d4;
  border-radius: 8px;
  padding: 8px;
  width: 100%;
`;

// !SECTION
// SECTION: Main

interface Props {
  name: string;
  placeholder: string;
  label: string;
}

const Input = memo(({ name, label, placeholder }: Props) => {
  const [field] = useField({
    name,
  });

  return (
    <Wrapper>
      <Label htmlFor={name}>{label}</Label>
      <BaseInput
        value={field.value}
        id={name}
        placeholder={placeholder}
        onChange={field.onChange}
        onBlur={field.onBlur}
      />
    </Wrapper>
  );
});

export default Input;
