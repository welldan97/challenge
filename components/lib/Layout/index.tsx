import { memo } from 'react';
import { Logo, Box } from '../../styles';

// SECTION: Main

interface Props {
  isShrunk?: boolean;
  children: React.ReactNode;
}

const Layout = memo(({ isShrunk, children }: Props) => (
  <>
    <Logo src="/Logo.svg" alt="Utrust" width="120" height="28" />
    <Box isShrunk={isShrunk} data-testid="wrapper">
      {children}
    </Box>
  </>
));

export default Layout;
