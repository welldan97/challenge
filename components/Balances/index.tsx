import { useQuery } from '@tanstack/react-query';
import { memo } from 'react';
import balanceApi from '../../api/balanceApi';
import { Balance } from '../../lib/Balance';
import Button from '../lib/Button';
import Layout from '../lib/Layout';
import { Footer, Header } from '../styles';
import BalancesTable from './BalancesTable';

// SECTION: Main

interface Props {
  balances: Balance[];
}

const Balances = memo(({ balances }: Props) => (
  <Layout>
    <Header>My Ethereum addresses</Header>
    <BalancesTable balances={balances} />
    <Footer>
      <span>Please copy the address from which you wish to send money.</span>
      <Button type="anchor" data-testid="next" href="/send">
        Next
      </Button>
    </Footer>
  </Layout>
));

// !SECTION
// SECTION Container

interface ContainerProps {
  balances?: Balance[];
}

export default memo(({ balances: initialBalances }: ContainerProps) => {
  const { data: balances } = useQuery(
    ['balance', 'getAll'],
    balanceApi.getAll,
    initialBalances && { initialData: initialBalances, staleTime: 1000 },
  );

  return <Balances balances={balances} />;
});
