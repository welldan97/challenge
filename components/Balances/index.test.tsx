import { render, waitFor, screen } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Balances from '.';
import balances from '../../lib/test/fixtures/balances';
import { Response } from '../../lib/Response';
import { Balance } from '../../lib/Balance';
import AppWrapper from '../lib/AppWrapper';

// SECTION: Main

describe('Balances', () => {
  let mock: MockAdapter;

  beforeEach(() => {
    mock = new MockAdapter(axios);

    mock.onGet('/api/balances').reply(200, {
      status: 'success',
      value: balances,
    } as Response<Balance[]>);
  });

  afterEach(() => {
    mock.reset();
  });

  test('render', async () => {
    const { container } = render(
      <AppWrapper>
        <Balances />
      </AppWrapper>,
    );

    await waitFor(() => {
      expect(screen.getByTestId('wrapper')).toBeTruthy();
    });

    expect(container).toMatchSnapshot();
  });
});
