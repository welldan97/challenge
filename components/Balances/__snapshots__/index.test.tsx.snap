// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balances render 1`] = `
<div>
  <div
    class="sc-4d880b92-0 fwseqv"
  >
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        alt="Utrust"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
  </div>
  <div
    class="sc-4d880b92-1 ccsBPG"
    data-testid="wrapper"
  >
    <header
      class="sc-4d880b92-2 jLouef"
    >
      My Ethereum addresses
    </header>
    <table
      class="sc-50244374-0 dsjiEy"
    >
      <tbody>
        <tr
          class="sc-50244374-1 cYIIdr"
        >
          <td>
            <span
              class="sc-50244374-2 bGcnud"
            >
              TEST_ADDRESS1
            </span>
          </td>
          <td
            class="sc-50244374-3 jGtUXk"
          >
            <span
              class="sc-50244374-4 jMsZsK"
            >
              TEST_VALUE1
            </span>
            <span
              class="sc-50244374-5 hbdULL"
            >
               ETH
            </span>
          </td>
        </tr>
        <tr
          class="sc-50244374-1 cYIIdr"
        >
          <td>
            <span
              class="sc-50244374-2 bGcnud"
            >
              TEST_ADDRESS2
            </span>
          </td>
          <td
            class="sc-50244374-3 jGtUXk"
          >
            <span
              class="sc-50244374-4 jMsZsK"
            >
              TEST_VALUE2
            </span>
            <span
              class="sc-50244374-5 hbdULL"
            >
               ETH
            </span>
          </td>
        </tr>
        <tr
          class="sc-50244374-1 cYIIdr"
        >
          <td>
            <span
              class="sc-50244374-2 bGcnud"
            >
              TEST_ADDRESS3
            </span>
          </td>
          <td
            class="sc-50244374-3 jGtUXk"
          >
            <span
              class="sc-50244374-4 jMsZsK"
            >
              TEST_VALUE3
            </span>
            <span
              class="sc-50244374-5 hbdULL"
            >
               ETH
            </span>
          </td>
        </tr>
      </tbody>
    </table>
    <footer
      class="sc-4d880b92-3 kNshHL"
    >
      <span>
        Please copy the address from which you wish to send money.
      </span>
      <a
        class="sc-23316aab-1 dzqFtz"
        data-testid="next"
        href="/send"
      >
        Next
      </a>
    </footer>
  </div>
</div>
`;
